### PR TITLE
Implement a backend for Quarkus Cache using Redis

### LINKED ISSUE #27785: Integrate Redis with the Quarkus Cache
### Description
Allows to use Redis instead of Caffeine with the Quarkus Cache extension.

### PR DESCRIPTION
**Pull request by cescoffier**
Closes #27785 

This PR extends Quarkus Cache to allow plugging various backend technology and provide an implementation for Redis.

CC @holly-cummins could be helpful.

### CODE REVIEW

**Comment by geoand** [comment]
Seems like the docs job failed

**Comment by cescoffier** [comment]
Will have a look in 5 min

**Comment by geoand** [comment]
@gsmet any idea?

**Comment by cescoffier** [comment]
I removed the usage of config mapping to use config roots and config groups. It seems to work (on my machine(tm))

**Comment by geoand** [comment]
Now we need to fix the BOM :(

**Comment by cescoffier** [comment]
@geoand let's retry....

**Comment by karesti** [comment]
@cescoffier @geoand happy to see this PR!!

**Comment by geoand** [comment]
@gwenneg could you have a look at this one please as would really like to have it in the next release if possible?

**Comment by gwenneg** [comment]
This is my top priority this morning, sorry for the delay.

File: extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheProcessor.java [hunk 1]
```diff
@@ -232,17 +233,26 @@ private List<Throwable> validateKeyGeneratorsDefaultConstructor(CombinedIndexBui
 
     @BuildStep
     @Record(RUNTIME_INIT)
-    SyntheticBeanBuildItem configureCacheManagerSyntheticBean(CacheNamesBuildItem cacheNames,
-            CacheManagerRecorder cacheManagerRecorder, Optional<MetricsCapabilityBuildItem> metricsCapability) {
+    void cacheManagerInfos(BuildProducer<CacheManagerInfoBuildItem> producer,
+            Optional<MetricsCapabilityBuildItem> metricsCapability, CacheManagerRecorder recorder) {
+        producer.produce(new CacheManagerInfoBuildItem(recorder.noOpCacheManagerInfo()));
+        producer.produce(new CacheManagerInfoBuildItem(recorder.getCacheManagerInfoWithoutMetrics()));
```
Comment (gwenneg) [inline extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheProcessor.java:239]: Is this needed when Micrometer is available?

**Review by gwenneg — COMMENTED** [review]
Am I the only one having issues loading the cache SPI modules as Maven projects in IntelliJ?

Here's a first round of comments. I'm still reviewing the Redis impl and tests.

**Comment by geoand** [comment]
> Am I the only one having issues loading the cache SPI modules as Maven projects in IntelliJ?

What kind of problems?

**Comment by geoand** [comment]
@cescoffier do you want to address the suggestions or should I?

File: extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheBuildConfig.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package io.quarkus.cache.runtime;
+
+import static io.quarkus.runtime.annotations.ConfigPhase.BUILD_AND_RUN_TIME_FIXED;
+
+import io.quarkus.runtime.annotations.ConfigRoot;
+import io.smallrye.config.ConfigMapping;
+import io.smallrye.config.WithDefault;
+
+@ConfigRoot(phase = BUILD_AND_RUN_TIME_FIXED)
```
[the remark this answers is not in the document]
Comment (geoand) [reply extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheBuildConfig.java:9]: I don't understand the concern here. Can you explain?

File: extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheProcessor.java (hunk 1 above)
Comment (geoand) [reply extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheProcessor.java:239]: We could certainly remove it, but it makes no real difference as the supplier is only invoked when necessary

**Comment by geoand** [comment]
Weird, I did not have that issue...

File: extensions/cache/deployment-spi/src/main/java/io/quarkus/cache/deployment/spi/CacheManagerInfoBuildItem.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package io.quarkus.cache.deployment.spi;
+
+import io.quarkus.builder.item.MultiBuildItem;
+import io.quarkus.cache.CacheManagerInfo;
+
+/**
+ * TODO: add
```
[the remark this answers is not in the document]
Comment (geoand) [reply extensions/cache/deployment-spi/src/main/java/io/quarkus/cache/deployment/spi/CacheManagerInfoBuildItem.java:7]: Yeah, we should probable add it :)

File: extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheBuildConfig.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (gwenneg) [reply extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheBuildConfig.java:9]: I found the answer in `RedisCacheProcessor`. `BUILD_AND_RUN_TIME_FIXED` makes sense now.

**Comment by cescoffier** [comment]
@geoand go ahead, I'm full of meetings.

File: bom/application/pom.xml [hunk 4]
```diff
@@ -5851,12 +5856,22 @@
                 <artifactId>quarkus-redis-client</artifactId>
                 <version>${project.version}</version>
             </dependency>
+            <dependency>
+                <groupId>io.quarkus</groupId>
+                <artifactId>quarkus-redis-cache</artifactId>
```
[the remark this answers is not in the document]
Comment (geoand) [reply bom/application/pom.xml:5861]: @gsmet WDYT?

File: docs/src/main/asciidoc/cache.adoc [hunk 5]
```diff
@@ -25,6 +25,13 @@ Since the weather forecast is updated once every twelve hours, caching the servi
 
 We'll do that using a single Quarkus annotation.
 
+[NOTE]
+====
+In this guide, we use the default Quarkus Cache backend (Caffeine).
```
[the remark this answers is not in the document]
Comment (geoand) [reply docs/src/main/asciidoc/cache.adoc:30]: Hm, I personally think backend is better here, but I don't really mind.

**Comment by geoand** [comment]
I fixed most of the comments. The only one I think we really need to resolve before merging is the name of the artifact

**Review by gwenneg — COMMENTED** [review]
Here's another round of minor comments.

File: extensions/redis-cache/runtime/src/main/java/io/quarkus/cache/redis/runtime/RedisCache.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+     * @param clazz the class of the value
+     * @param valueLoader the value loader called when there is no value stored in the cache
+     * @param <K> the type of key
+     * @param <V> the type of value
+     * @return the Uni emitting the cached value.
+     */
+    <K, V> Uni<V> get(K key, Class<V> clazz, Function<K, V> valueLoader);
+
+    /**
+     * Allows retrieving a value from the Redis cache.
```
[the remark this answers is not in the document]
Comment (geoand) [reply extensions/redis-cache/runtime/src/main/java/io/quarkus/cache/redis/runtime/RedisCache.java:55]: This is one for @cescoffier

File: extensions/redis-cache/runtime/src/main/java/io/quarkus/cache/redis/runtime/RedisCacheImpl.java [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+import io.vertx.mutiny.redis.client.Command;
+import io.vertx.mutiny.redis.client.Redis;
+import io.vertx.mutiny.redis.client.RedisConnection;
+import io.vertx.mutiny.redis.client.Request;
+
+/**
+ * This class is an internal Quarkus cache implementation using Redis.
+ * Do not use it explicitly from your Quarkus application.
+ */
+public class RedisCacheImpl<K, V> extends AbstractCache implements RedisCache {
```
[the remark this answers is not in the document]
Comment (geoand) [reply extensions/redis-cache/runtime/src/main/java/io/quarkus/cache/redis/runtime/RedisCacheImpl.java:32]: I removed them, but I kept a separate commit as I think it's important to have history of this

File: docs/src/main/asciidoc/cache.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (cescoffier) [reply docs/src/main/asciidoc/cache.adoc:30]: I believe it should be backend... everywhere (so, you both happy, right?)

File: bom/application/pom.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (cescoffier) [reply bom/application/pom.xml:5866]: This is a good question. To explain the current name, it's only to make things a bit more convenient as in the module list it's just after the Redis client extension. As they are closely related, it was a way to group them. But, I do not have a strong opinion.

File: docs/src/main/asciidoc/cache.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (geoand) [reply docs/src/main/asciidoc/cache.adoc:30]: I am :)

File: bom/application/pom.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (geoand) [reply bom/application/pom.xml:5866]: > But, I do not have a strong opinion.

Me neither

**Comment by geoand** [comment]
@gwenneg are you fine with getting this in?

**Review by gwenneg — APPROVED** [review]
> @gwenneg are you fine with getting this in and addressing any minor concerns that remain later?

Sure! LGTM

Thanks for this addition to quarkus-cache, that's really awesome!

**Comment by geoand** [comment]
Thank you! I'll wait until tomorrow in case there are any last objections

**Pull request merged**